rust/lance/src/index/vector/ivf/io.rs

```diff
@@ -36,11 +37,40 @@ use crate::Result;
 pub(super) async fn write_index_partitions(
     writer: &mut dyn Writer,
     ivf: &mut Ivf,
-    new_data: impl Stream<Item = DFResult<(Vec<ScalarValue>, Vec<RecordBatch>)>> + Unpin,
+    streams: Vec<impl Stream<Item = Result<RecordBatch>>>,
     existing_partitions: Option<&IVFIndex>,
 ) -> Result<()> {
-    let mut new_data = new_data.peekable();
-    let mut new_data_ref = Pin::new(&mut new_data);
+    // build the inital heap
+    let mut streams_heap = BinaryHeap::new();
+    let mut new_streams = vec![];
+
+    for stream in streams {
+        let mut stream = Box::pin(stream.peekable());
+
+        match stream.as_mut().peek().await {
```

Comment:
This is still memory bound, right?

Before you scan all data, you don't know whether the "first" partition is fully loaded.